public/app/features/dashboard/state/PanelQueryRunner.ts

```diff
@@ -0,0 +1,184 @@
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { Subject, Unsubscribable, PartialObserver } from 'rxjs';
+import {
+  guessFieldTypes,
+  toSeriesData,
+  PanelData,
+  LoadingState,
+  DataQuery,
+  TimeRange,
+  ScopedVars,
+  DataRequestInfo,
+  SeriesData,
+  LegacyResponseData,
+  DataQueryError,
+  toLegacyResponseData,
+  isSeriesData,
+  DataSourceApi,
+} from '@grafana/ui';
+
+import kbn from 'app/core/utils/kbn';
+
+/**
+ * Query response may come as events. any null|missing value just means
+ * that is unknown info in the request
+ */
+export interface PanelDataEvent extends Partial<PanelData> {
+  eventId: number;
+}
+
+export interface QueryRunnerOptions {
+  ds?: DataSourceApi; // if they already have the datasource, don't look it up
+  datasource: string | null;
+  queries: DataQuery[];
+  panelId: number;
+  dashboardId?: number;
+  timezone?: string;
+  timeRange?: TimeRange;
+  widthPixels: number;
+  minInterval?: string;
+  maxDataPoints?: number;
+  scopedVars?: ScopedVars;
+  cacheTimeout?: string;
+}
+
+export class PanelQueryRunner {
```

Comment:
Would it make sense to make this class, as well as `QueryRunnerOptions` a generic one, to make it more panel/dashboard concept independent? I think this could be used in i.e. Explore as well, and there the panelId/dashboardId are not used AFAIK.